### PR TITLE
Add OpenWebText and Organize

### PR DESCRIPTION
**Pull request by Quentin-Anthony**
@ndey96

### CODE REVIEW

**Comment by Quentin-Anthony** [comment]
Draft PR until I finish the README

File: mup_scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Experiment Reproduction
+
+Install requirements with `pip install -r requirements.txt`. We used the PyTorch NGC container for GPU-based runs, but any environment with the dependencies from [the main README](https://github.com/EleutherAI/nanoGPT-mup?tab=readme-ov-file#install) installed will suffice.
+
+
+# Coordinate Checks
+
+TODO
```
Comment (ndey96) [inline mup_scripts/README.md:8]: `bash mup_examples/coord_check_shakespeare_char/sp/run.sh`
...
`bash mup_examples/coord_check_shakespeare_char/mup/run.sh`

File: mup_scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+# Learning Rate Transfer
```
Comment (ndey96) [inline mup_scripts/README.md:10]: "muTransfer"

File: mup_scripts/README.md (hunk 1 above)
Comment (ndey96) [reply mup_scripts/README.md:8]: then run `plot.ipynb`

File: mup_scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+TODO
```
Comment (ndey96) [inline mup_scripts/README.md:12]: `bash mup_examples/mutransfer_lr_shakespeare_char/mup/run.sh`
`bash mup_examples/mutransfer_lr_shakespeare_char/sp/run.sh`
or
`bash mup_examples/mutransfer_lr_owt/mup/run.sh`
`bash mup_examples/mutransfer_lr_owt/sp/run.sh`

File: mup_scripts/mutransfer_lr/mup/run_openwebtext.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                --batch_size=4 \
+                --block_size=1024 \
+                --n_layer=2 \
+                --n_head=$n_heads \
+                --n_embd=$width \
+                --dropout=0.0 \
+                --bias=False \
+                --init_std=0.02 \
+                --learning_rate=$lr \
+                --max_iters=122 \
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/mup/run_openwebtext.sh:42]: @Quentin-Anthony would it be possible for you to rerun with 1000 iters and global batch size of 32? This would make things much less noisy. I only selected 122 iters because I was running on my macbook. Hopefully then we wouldn't need to plot "mean train loss" which is a little gross. Then this test has marginal value above the shakespeare_char test

File: mup_scripts/mutransfer_lr/plot.ipynb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    "        self.scalarMap = cm.ScalarMappable(norm=self.norm, cmap=self.cmap)\n",
+    "\n",
+    "    def get_rgb(self, val):\n",
+    "        return self.scalarMap.to_rgba(val)\n",
+    "\n",
+    "\n",
+    "color_helper = MplColorHelper('viridis', 0, len(widths)-1)\n",
+    "n_cols = 2#len(parameterizations)\n",
+    "n_rows = 1\n",
+    "fig, axes = plt.subplots(n_rows, n_cols, figsize=(8*n_cols, 6*n_rows))\n",
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/plot.ipynb:71]: `fig, axes = plt.subplots(n_rows, n_cols, figsize=(4*n_cols, 3.33*n_rows))` to match the other plot

File: mup_scripts/mutransfer_lr/plot.ipynb [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    "# axes = axes.reshape((n_rows, n_cols))\n",
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/plot.ipynb:72]: delete

File: mup_scripts/mutransfer_lr/plot.ipynb [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    "        sem_losses = []\n",
+    "        lrs_to_plot = []\n",
+    "        for lr in lrs:\n",
+    "            losses = []\n",
+    "            for seed in seeds:\n",
+    "                job_name = f'width{width}_depth2_seed{seed}_lr{lr:.20f}'.rstrip('0')\n",
+    "                csv_path = os.path.join(parameterization, 'out', job_name, 'log.csv')\n",
+    "                if os.path.exists(csv_path):\n",
+    "                    ckpt_df = pd.read_csv(csv_path)\n",
+    "                    losses.append(ckpt_df['train/loss'].mean())\n",
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/plot.ipynb:89]: Ideally we plot something like
`losses.append(ckpt_df['train/loss'].min())` or `losses.append(ckpt_df['train/loss'].ewm(alpha=0.9).mean().values[-1])` instead once the OWT test has more steps and is larger scale

File: mup_scripts/mutransfer_lr/mup/run_openwebtext.sh (hunk 4 above)
Comment (ndey96) [reply mup_scripts/mutransfer_lr/mup/run_openwebtext.sh:42]: Can you also re-enable LR decay to 10x less than the maximum? I only chose constant because shakespeare_char was so small scale and short

File: mup_scripts/mutransfer_lr/sp/run_openwebtext.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+for width in 256 512 1024 2048
+do
+    # for lr in 0.125 0.0625 0.03125 0.015625 0.0078125 0.00390625 0.001953125 0.0009765625 0.00048828125 0.000244140625 0.0001220703125 0.00006103515625
+    for lr in 0.00390625 0.001953125 0.0009765625 0.00048828125 0.000244140625 0.0001220703125 0.00006103515625 0.00003051757812
+    do
+        for seed in 1 2 3
+        do
+            head_size=64
+            n_heads=$((width / head_size))
+            out_dir="mutransfer_lr/sp/out/width${width}_depth2_seed${seed}_lr${lr}"
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/sp/run_openwebtext.sh:17]: instead please move to a dir called `mutransfer_lr_owt` to go alongside `mutransfer_lr_shakespeare_char`

File: mup_scripts/mutransfer_lr/plot.ipynb [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    "            optimal_lrs.append(lrs_to_plot[optimum_idx])\n",
+    "            optimal_losses.append(mean_losses[optimum_idx])\n",
+    "        \n",
+    "    ax.plot(optimal_lrs, optimal_losses, color='red', linestyle='none', marker='o')\n",
+    "    ax.set_xscale('log', base=2)\n",
+    "    ax.set_xlabel('Learning rate')\n",
+    "    ax.set_title(parameterization_str)\n",
+    "    ax.set_ylim(7, 9)\n",
+    "\n",
+    "axes[0].legend(title='Width')\n",
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/plot.ipynb:114]: Maybe `axes[1].legend(title='Width')` will look better since the muP curves leave some negative space

File: mup_scripts/mutransfer_lr/plot.ipynb [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    "# axes[0].set_ylabel('Mean Train loss\\nshakespeare_char')\n",
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/plot.ipynb:115]: Our axis label should be truthful

File: mup_scripts/mutransfer_lr/plot.ipynb [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    "                    #print(f'Missing {csv_path}')\n",
+    "            if len(losses):\n",
+    "                mean_losses.append(np.mean(losses))\n",
+    "                sem_losses.append(np.std(losses, ddof=1) / np.sqrt(len(losses)))\n",
+    "                lrs_to_plot.append(lr)\n",
+    "        \n",
+    "        mean_losses = np.array(mean_losses)\n",
+    "        sem_losses = np.array(sem_losses)\n",
+    "        ax.plot(lrs_to_plot, mean_losses, label=width, marker='o', color=color_helper.get_rgb(width_idx))\n",
+    "        ax.fill_between(lrs_to_plot, mean_losses-sem_losses, mean_losses+sem_losses, color=color_helper.get_rgb(width_idx), alpha=0.5)\n",
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/plot.ipynb:101]: `alpha=0.33` looks a little nicer

File: mup_scripts/mutransfer_lr/sp/run_openwebtext.sh [hunk 12]
```diff
@@ -0,0 +1,54 @@
+# Single-GPU Launching
+LAUNCHER=python
+
+# Multi-GPU Launching (single node)
+#GPU=2
+#LAUNCHER=torchrun --standalone --nproc_per_node=$GPU
+
+for width in 256 512 1024 2048
+do
+    # for lr in 0.125 0.0625 0.03125 0.015625 0.0078125 0.00390625 0.001953125 0.0009765625 0.00048828125 0.000244140625 0.0001220703125 0.00006103515625
```
Comment (ndey96) [inline mup_scripts/mutransfer_lr/sp/run_openwebtext.sh:10]: delete if unused

File: mup_scripts/mutransfer_lr/plot.ipynb (hunk 10 above)
Comment (ndey96) [reply mup_scripts/mutransfer_lr/plot.ipynb:115]: Would be nice to label as `Train loss on OpenWebText`

**Review by ndey96 — CHANGES_REQUESTED** [review]
Thanks for adding this @Quentin-Anthony ! Just suggested several small changes to make things more consistent between our branches. I also refactored my branch's folder structure to be
```
mup_examples/
    coord_check_shakespeare_char/
    mutransfer_lr_shakespeare_char/
    TODO: mutransfer_lr_owt/
```
Can you follow this structure and ensure there are no merge conflicts?

Thanks!

File: mup_scripts/mutransfer_lr/mup/run_openwebtext.sh (hunk 4 above)
Comment (Quentin-Anthony) [reply mup_scripts/mutransfer_lr/mup/run_openwebtext.sh:42]: Running these now

Comment (Quentin-Anthony) [reply mup_scripts/mutransfer_lr/mup/run_openwebtext.sh]: image

File: mup_scripts/mutransfer_lr/plot.ipynb (hunk 9 above)
Comment (Quentin-Anthony) [reply mup_scripts/mutransfer_lr/plot.ipynb]: Looks better staying as `axes[0].legend(title='Width')` with the new figures

File: mup_scripts/mutransfer_lr/plot.ipynb (hunk 7 above)
Comment (Quentin-Anthony) [reply mup_scripts/mutransfer_lr/plot.ipynb]: went with `losses.append(ckpt_df['train/loss'].ewm(alpha=0.9).mean().values[-1])`

**Comment by ndey96** [comment]
@Quentin-Anthony looking great now! Can you just undo the changes to the `mutransfer_lr_shakespeare_char` folder? Then I'm good to merge

**Comment by Quentin-Anthony** [comment]
> @Quentin-Anthony looking great now! Can you just undo the changes to the `mutransfer_lr_shakespeare_char` folder? Then I'm good to merge

Done!
